modules/stochastic_tools/src/surrogates/LibtorchSimpleNNTrainer.C

```diff
@@ -0,0 +1,199 @@
+    _sampler_row(getSamplerData()),
+    _response(getTrainingData<Real>(getParam<ReporterName>("response"))),
+    _num_batches(getParam<unsigned int>("num_batches")),
+    _num_epocs(getParam<unsigned int>("num_epochs")),
+    _rel_loss_tol(getParam<Real>("rel_loss_tol")),
+    _num_neurons_per_layer(declareModelData<std::vector<unsigned int>>(
+        "num_neurons_per_layer", getParam<std::vector<unsigned int>>("num_neurons_per_layer"))),
+    _num_hidden_layers(
+        declareModelData<unsigned int>("num_hidden_layers", _num_neurons_per_layer.size())),
+    _filename(getParam<std::string>("filename")),
+    _read_from_file(getParam<bool>("read_from_file")),
+    _learning_rate(getParam<Real>("learning_rate")),
+    _print_epoch_loss(getParam<unsigned int>("print_epoch_loss"))
+#ifdef LIBTORCH_ENABLED
+    ,
+    _nn(declareModelData<std::shared_ptr<StochasticTools::LibtorchSimpleNeuralNet>>("nn"))
+#endif
+{
+  // We check if MOOSE is compiled with torch, if not this throws an error
+  StochasticToolsApp::requiresTorch(*this);
+
+#ifdef LIBTORCH_ENABLED
+  // Fixing the RNG seed to make sure every experiment is the same.
+  // Otherwise sampling / stochastic gradient descent would be different.
+  torch::manual_seed(getParam<unsigned int>("seed"));
+#endif
+}
+
+void
+LibtorchSimpleNNTrainer::preTrain()
+{
+  // Resize to number of sample points
+  _flattened_data.resize(_sampler.getNumberOfLocalRows() * _sampler.getNumberOfCols());
+  _flattened_response.resize(_sampler.getNumberOfLocalRows());
+}
+
+void
+LibtorchSimpleNNTrainer::train()
+{
+  unsigned int num_parameters = _sampler.getNumberOfCols();
+
+  for (unsigned int i = 0; i < num_parameters; ++i)
+    _flattened_data[_local_row * num_parameters + i] = _sampler_row[i];
+
+  _flattened_response[_local_row] = _response;
+}
+
+void
+LibtorchSimpleNNTrainer::postTrain()
+{
+  _communicator.allgather(_flattened_data);
+  _communicator.allgather(_flattened_response);
+
+#ifdef LIBTORCH_ENABLED
+
+  // Then, we create and load our Tensors
+  unsigned int n_rows = _flattened_response.size();
+  unsigned int n_cols = _sampler.getNumberOfCols();
+
+  // The default data type in pytorch is float, while we use double in MOOSE.
+  // Therefore, in some cases we have to convert Tensors to double.
+  auto options = torch::TensorOptions().dtype(at::kDouble);
+  torch::Tensor data_tensor =
+      torch::from_blob(_flattened_data.data(), {n_rows, n_cols}, options).to(at::kDouble);
+  torch::Tensor response_tensor =
+      torch::from_blob(_flattened_response.data(), {n_rows}, options).to(at::kDouble);
+
+  // We create a custom data loader which can be used to select samples for the in
+  // the training process. See the header file for the definition of this structure.
+  StochasticTools::LibtorchDataset my_data(data_tensor, response_tensor);
+
+  // We initialize a data_loader for the training part.
+  unsigned int sample_per_batch = n_rows > _num_batches ? n_rows / _num_batches : 1;
+
```

Comment:
OK, right,  Rename `n_rows` to `n_samples` seems consistent with `sample_per_batch`